legion/engine/physics/data/capsule_controller_data.hpp

```diff
@@ -0,0 +1,134 @@
+            m_currentDisplacement += displacement;
+        }
+
+        L_ALWAYS_INLINE void setRadius(float radius)
+        {
+            m_modificationFlags.set(capsule_character_flag::cc_set_radius, true);
+            m_radius = radius;
+        }
+
+        L_ALWAYS_INLINE float getRadius() const noexcept { return m_radius; }
+
+        L_ALWAYS_INLINE void setHeight(float height)
+        {
+            m_modificationFlags.set(capsule_character_flag::cc_set_height, true);
+            m_height = height;
+        }
+
+        L_ALWAYS_INLINE float getHeight() const noexcept { return m_height; }
+
+        L_ALWAYS_INLINE const math::vec3& getCurrentDisplacement() const noexcept { return m_currentDisplacement; }
+
+        L_ALWAYS_INLINE void resetDisplacement() { m_currentDisplacement = math::vec3(0); }
+
+        L_ALWAYS_INLINE const std::bitset<capsule_character_flag::cc_max>& getModificationFlags() const noexcept { return m_modificationFlags; }
+
+        L_ALWAYS_INLINE void resetModificationFlags() { m_modificationFlags.reset(); }
+
+        template<class preset>
+        pointer<const preset> getReadOnlyPreset() const
+        {
+            static size_type searchHash = typeHash<preset>();
+
+            for (const controller_preset& contPreset : m_presets)
+            {
+                if (searchHash == contPreset.hash)
+                {
+                    const void* specificsPtr = &contPreset.specifics;
+                    return { static_cast<const preset*>(specificsPtr) };
+                }
+            }
+
+            return { nullptr };
+        }
+
+        template<class preset>
+        pointer<preset> getPreset() 
+        {
+            static size_type searchHash = typeHash<preset>();
+
+            for (controller_preset& contPreset : m_presets)
+            {
+                if (searchHash == contPreset.hash)
+                {
+                    void* specificsPtr = &contPreset.specifics;
+                    return { static_cast<preset*>(specificsPtr) };
+                }
+            }
+
+            return { nullptr };
+        }
+
+        template<class preset>
+        pointer<preset> addPreset(preset& presetSpecifics)
+        {
+            static_assert(std::is_standard_layout<preset>());
+            static size_type searchHash = typeHash<preset>();
+
+            for (controller_preset& contPreset : m_presets)
+            {
+                if (searchHash == contPreset.hash)
+                {
+                    void* specificsPtr = &contPreset.specifics;
+                    return { static_cast<preset*>(specificsPtr) };
+                }
+            }
+
+            controller_preset newPreset;
+            newPreset.hash = searchHash;
+            //memcpy here would be fine since preset is of standard layout
+            memcpy(&newPreset.specifics, &presetSpecifics, sizeof(preset)); 
+
+            m_presets.push_back(newPreset);
```

Comment:
furthermore, since you're doing allocations here anyways for the pushback. do realise that putting both presets in a union means that every item in the m_presets vector takes up as much memory as the largest preset specific. depending on how often these presets rly get accessed, and how many there will exist at any point in time, it might be worth having a polymorphic pointer instead of a union.

Reply:
we do iterate through them at every physics time step. The presets represent features of a character controller so it would make sense to say each character controller would have <10 presets. I don't currently see the benefit of making it polymorphic since this would just add an extra indirection everytime we iterate through a preset